src/se/vidstige/jadb/JadbConnection.java

```diff
@@ -38,9 +39,35 @@ public void getHostVersion() throws IOException, JadbException {
         devices.send("host:devices");
         devices.verifyResponse();
         String body = devices.readString();
+        devices.close();
         return parseDevices(body);
     }
 
+    public AsyncActionHandler watchDevices(final DeviceDetectionListener listener) throws IOException, JadbException {
+        final Transport transport = createTransport();
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    transport.send("host:track-devices");
+                    transport.verifyResponse();
+                    boolean r = false;
+                    do {
+                        List<JadbDevice> list = parseDevices(transport.readString());
+                        r = listener.detect(list);
+                    } while (r);
+                } catch (SocketException e) {
+                    // socket closed from another thread
+                } catch (Exception e) {
+                    Thread t = Thread.currentThread();
+                    t.getUncaughtExceptionHandler().uncaughtException(t, e);
+                }
+            }
+        }).start();
+
+        return new AsyncActionHandler(transport);
+    }
```

Comment:
Yes, that's true. Perhaps it's best to just let the returned object have a blocking `run()` method as well as the `stop()` method. This way it's super easy to plug into a thread. 

```java
private class DeviceWatcher implements Runnable {
    private final Transport transport;
    private final DeviceDetectionListener listener;
    private boolean stopped;

    public DeviceWatcher(Transport transport, final DeviceDetectionListener listener) {
        this.transport = transport;
        this.listener = listerner;
        this.stopped = false;
    }

    public void run() {
        try {
            while (true) {
                List<JadbDevice> list = parseDevices(transport.readString());
                for (JadbDevice device : list) {
                    listener.onDevice(device);
                }
            }
        }
        catch (IOException io) {
            if (!stopped) {
                listerner.onException(io);
            }
        }
    }

    public void stop() throws IOException {
        stopped = true;
        transport.close();
    }
}

public DeviceWatcher createWatcher(final DeviceDetectionListener listener) throws IOException, JadbException {
    final Transport transport = createTransport();
    transport.send("host:track-devices");
    transport.verifyResponse();
    return new DeviceWatcher(listener);
}
```